apps/api/src/campaign/campaign.service.ts

```diff
@@ -450,95 +450,71 @@ export class CampaignService {
-          personId: paymentData.personId,
-          chargedAmount: paymentData.chargedAmount,
-          extPaymentMethodId: 'subscription',
-        },
-        select: donationNotificationSelect,
-      })
-
-      if (donation && newDonationStatus == DonationStatus.succeeded) {
-        donation.status = newDonationStatus
-        this.notificationService.sendNotification('successfulDonation', donation)
+    //Update existing donation or create new in a transaction that
+    //also increments the vault amount and marks campaign as completed
+    //if target amount is reached
+    return await this.prisma.$transaction(async (tx) => {
+      let donationId
+      // Find donation by extPaymentIntentId
+      const existingDonation = await this.findExistingDonation(tx, paymentData)
+
+      //if missing create the donation with the incoming status
+      if (!existingDonation) {
+        const newDonation = await this.createIncomingDonation(
+          tx,
+          paymentData,
+          newDonationStatus,
+          campaign,
+        )
+        donationId = newDonation.id
       }
-
-      Logger.debug('Donation found by subscription: ', donation)
-    }
-
-    //if missing create the donation with the incoming status
-    if (!donation) {
-      Logger.debug(
-        'No donation exists with extPaymentIntentId: ' +
-          paymentData.paymentIntentId +
-          ' Creating new donation with status: ' +
+      //donation exists, so check if it is safe to update it
+      else {
+        const updatedDonation = await this.updateDonationIfAllowed(
+          tx,
+          existingDonation,
           newDonationStatus,
-      )
+          paymentData,
+        )
+        donationId = updatedDonation?.id
+      }
 
-      try {
-        donation = await this.prisma.donation.create({
-          data: {
-            amount: paymentData.netAmount,
-            chargedAmount: paymentData.chargedAmount,
-            currency: campaign.currency,
-            targetVault: targetVaultData,
-            provider: paymentData.paymentProvider,
-            type: DonationType.donation,
-            status: newDonationStatus,
-            extCustomerId: paymentData.stripeCustomerId ?? '',
-            extPaymentIntentId: paymentData.paymentIntentId,
-            extPaymentMethodId: paymentData.paymentMethodId ?? '',
-            billingName: paymentData.billingName,
-            billingEmail: paymentData.billingEmail,
-            person: paymentData.personId ? { connect: { id: paymentData.personId } } : {},
-          },
-          select: donationNotificationSelect,
-        })
-        if (newDonationStatus === DonationStatus.succeeded) {
-          this.notificationService.sendNotification('successfulDonation', donation)
+      //For successful donations we will also need to link them to user if not marked as anonymous
+      if (donationId && newDonationStatus === DonationStatus.succeeded) {
+        if (metadata?.isAnonymous != 'true') {
```

Comment:
!==   - just to be consistent and sure that equality will work as expected

Reply:
done